pam/internal/adapter/gdmmodel.go

```diff
@@ -271,3 +289,25 @@ func (m gdmModel) changeStage(s proto.Stage) tea.Cmd {
 		return nil
 	}
 }
+
+func (m gdmModel) stopConversations() gdmModel {
+	// We're about to exit: let's ensure that all the messages have been processed.
+
+	time.Sleep(gdmPollFrequency * 2)
+
+	wait := make(chan struct{})
+	go func() {
+		for gdm.ConversationInProgress() {
```

Comment:
This is still a busy loop, no? it will spin instead of waiting for a number of mutex release?

Reply:
Oh well... Indeed it is, but I think we need to wait here, and per se a gdm conversation can be initiated by multiple actors.

I can refactor this though, ideally I wanted to keep as simple as possible since this happens when we're exiting only and we don't have really resources constraints.

Reply:
So something like...

```diff
diff --git a/pam/internal/adapter/gdmmodel.go b/pam/internal/adapter/gdmmodel.go
index 451612f..20fda2f 100644
--- a/pam/internal/adapter/gdmmodel.go
+++ b/pam/internal/adapter/gdmmodel.go
@@ -297,8 +297,7 @@ func (m gdmModel) stopConversations() gdmModel {
 
 	wait := make(chan struct{})
 	go func() {
-		for gdm.ConversationInProgress() {
-		}
+		gdm.WaitForConversationsStopped()
 		wait <- struct{}{}
 	}()
 
diff --git a/pam/internal/gdm/conversation.go b/pam/internal/gdm/conversation.go
index a32bc5c..e2f8172 100644
--- a/pam/internal/gdm/conversation.go
+++ b/pam/internal/gdm/conversation.go
@@ -6,27 +6,49 @@ import (
 	"context"
 	"errors"
 	"fmt"
-	"sync/atomic"
+	"sync"
 
 	"github.com/msteinert/pam/v2"
 	"github.com/ubuntu/authd/internal/log"
 )
 
-var conversations atomic.Int32
+var conversations int
+var conversationsMu = sync.Mutex{}
+var conversationsCond = sync.Cond{L: &conversationsMu}
 
 // ConversationInProgress checks if conversations are currently active.
 func ConversationInProgress() bool {
-	return conversations.Load() > 0
+	return false
+	// return conversations.Load() > 0
+}
+
+func WaitForConversationsStopped() {
+	conversationsMu.Lock()
+	defer conversationsMu.Unlock()
+
+	for conversations != 0 {
+		conversationsCond.Wait()
+	}
 }
 
 func sendToGdm(pamMTx pam.ModuleTransaction, data []byte) ([]byte, error) {
-	conversations.Add(1)
-	defer conversations.Add(-1)
 	binReq, err := NewBinaryJSONProtoRequest(data)
 	if err != nil {
 		return nil, err
 	}
 	defer binReq.Release()
+
+	conversationsMu.Lock()
+	conversations++
+	conversationsMu.Unlock()
+
+	defer func() {
+		conversationsMu.Lock()
+		conversations--
+		conversationsCond.Broadcast()
+		conversationsMu.Unlock()
+	}()
+
 	res, err := pamMTx.StartConv(binReq)
 	if err != nil {
 		return nil, err
```

It seems to me that it adds lot of complexity to an operation that happens a lot during runtime (e.g. all the polling...) while spin-locking it's an one time thing only during shutdown...

Reply:
Ok, pushed the version we agree'd in pvt.